core/plugins/org.polarsys.capella.core.sirius.analysis/src/org/polarsys/capella/core/sirius/analysis/EventContextServices.java

```diff
@@ -0,0 +1,414 @@
+        && ((StateFragment) timeLapse.get()).getFinish() == end) {
+      result.add(new EventContext(ancestors.peek(), timeLapse.get(), false, ancestors.size() + 1));
+    }
+  }
+
+  /**
+   * Compute EventStructure for ExecutionEnd.
+   * 
+   * @param ancestors
+   *          Deque<EObject>
+   * @param result
+   *          List<EventContext>
+   * @param end
+   *          ExecutionEnd
+   */
+  private static void visit(Deque<CapellaElement> ancestors, List<EventContext> result, ExecutionEnd end,
+      Optional<TimeLapse> timeLapse) {
+    // Should not happen in current implementation of Capella Sequence diagrams
+    if (timeLapse.isPresent() && timeLapse.get().getStart() == end) {
+      result.add(new EventContext(ancestors.peek(), timeLapse.get(), true, ancestors.size() + 1));
+      ancestors.push(timeLapse.get());
+    }
+
+    // Execution without return branch end / ASyncCall
+    if (timeLapse.isPresent() && timeLapse.get().getFinish() == end) {
+      ancestors.pop();
+      result.add(new EventContext(ancestors.peek(), timeLapse.get(), false, ancestors.size() + 1));
+    }
+  }
+
+  /**
+   * Return covered instance role.
+   * 
+   * @param end
+   *          InteractionFragment
+   * @return covered instance role.
+   */
+  private static InstanceRole getCoveredInstanceRole(InteractionFragment end) {
+    InstanceRole covered = null;
+    if (end instanceof AbstractEnd) {
+      covered = getInteractionCache(AbstractEnd::getCovered, (AbstractEnd) end);
+    } else if (end instanceof InteractionState) {
+      covered = getInteractionCache(InteractionState::getCovered, (InteractionState) end);
+    }
+    return covered;
+  }
+
+  /**
+   * Compute cache for TimeLapse start and finish ends.
+   * 
+   * @param scenario
+   *          Scenario
+   */
+  private static void computeTimeLapseStartAndEndCache(Scenario scenario) {
+    // cache Execution.start -> Execution and Execution.finish -> Execution
+    // scan timelapse only one time
+    if (getInteractionFragmentToTimeLapseCache().isEmpty()) {
+      Stream<TimeLapse> timeLapses = scenario.getOwnedTimeLapses().stream()
+          .filter(tl -> !(tl instanceof AbstractFragment));
+      timeLapses.forEach(e -> {
+        // Keep Interaction to TimeLapse info in cache
+        // M2 allows to retrieve an Execution from an ExecutionEnd (execution.getFinish() when there is no
+        // return branch : ASyncCall)
+        // but not an Execution from a MessagEnd (SyncCall and ASyncCall) or a StateFragment from an
+        // InteractionState.
+        putTimeLapseInCache(e.getStart(), e);
+        putTimeLapseInCache(e.getFinish(), e);
+      });
+    }
+  }
+
+  /**
+   * Helper class to keep track of who "contains" who depending on the interleaving of the start/finish ends.
+   * 
+   * @author pcdavid
+   */
+  public static final class EventContext {
+    private final EObject parent;
+
+    private final boolean start;
+
+    private final EObject element;
```

Comment:
change to `child`

Reply:
I am not sure, this is really the current element for which we want the parent. 
Child would be used if we wanted to compute the children of element.